File: backend/src/main/java/com/furios/labirinto_dungeonico/dungeon/Room.java
package com.furios.labirinto_dungeonico.dungeon;

import com.fasterxml.jackson.annotation.JsonAutoDetect;
import com.furios.labirinto_dungeonico.combat.Enemy;
import com.furios.labirinto_dungeonico.item.Item;

import java.util.ArrayList;
import java.util.List;

@JsonAutoDetect(fieldVisibility = JsonAutoDetect.Visibility.ANY)
public class Room {

    private final String id;
    private final RoomType type;
    private final int x;
    private final int y;
    private final List<Item> items = new ArrayList<>();
    private final List<String> connectedRoomIds = new ArrayList<>();
    private Enemy enemy;

    public Room(String id, RoomType type, int x, int y) {
        this.id = id;
        this.type = type;
        this.x = x;
        this.y = y;
    }

    public String id() {
        return id;
    }

    public RoomType type() {
        return type;
    }

    /** Coluna da célula na grade do autômato celular (0-indexada, cresce para a direita). */
    public int x() {
        return x;
    }

    /** Linha da célula na grade do autômato celular (0-indexada, cresce para baixo). */
    public int y() {
        return y;
    }

    public List<Item> items() {
        return items;
    }

    /**
     * Ids das salas andáveis ortogonalmente adjacentes (cima/baixo/esquerda/direita) nesta
     * grade. Calculado pelo MapGenerator após a suavização do autômato celular; vazio para
     * salas do tipo WALL, que nunca aparecem como vizinhas de uma sala andável (RN-15).
     */
    public List<String> connectedRoomIds() {
        return connectedRoomIds;
    }

    public Enemy enemy() {
        return enemy;
    }

    public void setEnemy(Enemy enemy) {
        this.enemy = enemy;
    }
}

File: app/lib/core/models/room.dart
import 'enemy.dart';
import 'item.dart';

enum RoomType { entrance, empty, loot, enemy, exit, wall }

/// Uma célula da grade do autômato celular. Toda posição (x, y) do
/// [DungeonMap] tem uma [Room] correspondente, inclusive as do tipo
/// [RoomType.wall] — não há "buracos" na grade.
class Room {
  const Room({
    required this.id,
    required this.type,
    required this.x,
    required this.y,
    required this.items,
    required this.connectedRoomIds,
    required this.enemy,
  });

  factory Room.fromJson(Map<String, dynamic> json) => Room(
        id: json['id'] as String,
        type: RoomType.values.byName((json['type'] as String).toLowerCase()),
        x: json['x'] as int,
        y: json['y'] as int,
        items: (json['items'] as List)
            .map((e) => Item.fromJson(e as Map<String, dynamic>))
            .toList(),
        connectedRoomIds: (json['connectedRoomIds'] as List)
            .map((e) => e as String)
            .toList(),
        enemy: json['enemy'] == null
            ? null
            : Enemy.fromJson(json['enemy'] as Map<String, dynamic>),
      );

  final String id;
  final RoomType type;

  /// Coluna da célula na grade (0-indexada, cresce para a direita).
  final int x;

  /// Linha da célula na grade (0-indexada, cresce para baixo).
  final int y;
  final List<Item> items;

  /// Ids das salas andáveis ortogonalmente adjacentes. Sempre vazio para
  /// salas [RoomType.wall].
  final List<String> connectedRoomIds;

  /// Inimigo vivo na sala, ou `null`. Importante: `type` continua
  /// [RoomType.enemy] mesmo depois do inimigo derrotado (sem respawn nesta
  /// versão) — é este campo, não o `type`, que diz se ainda há combate
  /// pendente.
  final Enemy? enemy;
}
